### PR TITLE
[Compression - 4] Add decompressor interface

### PR DESCRIPTION
**Pull request by piraka9011**
Add the `BaseDecompressorInterface` that will be used to implement decompressors.

Signed-off-by: Anas Abou Allaban allabana@amazon.com

### CODE REVIEW

File: rosbag2_compression/include/rosbag2_compression/base_compressor_interface.hpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+// limitations under the License.
+
+#ifndef ROSBAG2_COMPRESSION__BASE_COMPRESSOR_INTERFACE_HPP_
+#define ROSBAG2_COMPRESSION__BASE_COMPRESSOR_INTERFACE_HPP_
+
+#include <string>
+
+#include "rosbag2_storage/serialized_bag_message.hpp"
+#include "rosbag2_compression/visibility_control.hpp"
+
```
Comment (Karsten1987) [inline rosbag2_compression/include/rosbag2_compression/base_compressor_interface.hpp]:
```suggestion
```

File: rosbag2_compression/include/rosbag2_compression/base_compressor_interface.hpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_COMPRESSION__BASE_COMPRESSOR_INTERFACE_HPP_
+#define ROSBAG2_COMPRESSION__BASE_COMPRESSOR_INTERFACE_HPP_
+
+#include <string>
+
+#include "rosbag2_storage/serialized_bag_message.hpp"
+#include "rosbag2_compression/visibility_control.hpp"
```
Comment (Karsten1987) [inline rosbag2_compression/include/rosbag2_compression/base_compressor_interface.hpp]: alpha

File: rosbag2_compression/test/rosbag2_compression/fake_decompressor.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "fake_decompressor.hpp"
```
Comment (Karsten1987) [inline rosbag2_compression/test/rosbag2_compression/fake_decompressor.cpp]: same comment as in #218 we might keep the mock interfaces.

**Comment by piraka9011** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8752/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4581/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7154/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8679/)

File: rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+public:
+  virtual ~BaseDecompressorInterface() = default;
+
+  /**
+   * Decompress a file on disk.
+   *
+   * \param uri Input file to decompress with file extension.
+   * \return The relative path to the decompressed file without the compressed extension.
+   */
+  virtual std::string decompress_uri(const std::string & uri) = 0;
```
Comment (thomas-moulard) [inline rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp]:
```suggestion
  virtual std::string decompress_uri(const std::string & uri) const = 0;
```
not sure - but should it be const?

File: rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * Decompress the serialized_data of a serialized bag message in place.
+   *
+   * \param bag_message A serialized bag message.
```
Comment (thomas-moulard) [inline rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp]:
```suggestion
   * \param[in,out] bag_message A serialized bag message.
```

http://www.doxygen.nl/manual/commands.html#cmdparam

File: rosbag2_compression/test/rosbag2_compression/fake_decompressor.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+std::string FakeDecompressor::decompress_uri(const std::string & uri)
+{
+  auto uri_path = rcpputils::fs::path(uri);
+  auto decompressed_path = rcpputils::fs::remove_extension(uri_path);
```
Comment (thomas-moulard) [inline rosbag2_compression/test/rosbag2_compression/fake_decompressor.cpp]: add const whenever possible

File: rosbag2_compression/test/rosbag2_compression/test_decompressor.cpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+public:
+  FakeDecompressor decompressor;
+};
+
+TEST_F(FakeDecompressorTest, test_compress_file_method)
+{
+  rcpputils::fs::path test_path("/path/file.txt");
+  auto compressed_uri = test_path.string() + "." + decompressor.get_decompression_identifier();
+  auto expected_decompressed_uri = test_path.string();
+  auto decompressed_uri = decompressor.decompress_uri(compressed_uri);
```
Comment (thomas-moulard) [inline rosbag2_compression/test/rosbag2_compression/test_decompressor.cpp]: const

File: rosbag2_compression/test/rosbag2_compression/test_decompressor.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  EXPECT_EQ(decompressed_uri, expected_decompressed_uri);
+}
+
+TEST_F(FakeDecompressorTest, test_compress_bag_method)
+{
+  auto bag_message = std::make_shared<rosbag2_storage::SerializedBagMessage>();
```
Comment (thomas-moulard) [inline rosbag2_compression/test/rosbag2_compression/test_decompressor.cpp]:
```suggestion
  const auto bag_message = std::make_shared<rosbag2_storage::SerializedBagMessage>();
```

File: rosbag2_compression/test/rosbag2_compression/fake_decompressor.cpp (hunk 3 above)
Comment (piraka9011) [reply rosbag2_compression/test/rosbag2_compression/fake_decompressor.cpp]: I didn't get a response on this from the previous PR but reiterating here:

This is technically a fake implementation and not a mock. I'm concerned of potential confusion down the line if we decide to call it a mock instead.
I will be making mocks for these down the line tho.

File: rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp (hunk 4 above)
Comment (piraka9011) [reply rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp]: There might be a compression implementation where someone changes member variables in this function so I would leave it as is.

Comment (thomas-moulard) [reply rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp]: :+1:

File: rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp (hunk 5 above)
Comment (piraka9011) [reply rosbag2_compression/include/rosbag2_compression/base_decompressor_interface.hpp]: This is helpful!

**Comment by piraka9011** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8773/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4599/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7175/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8697/)

**Comment by piraka9011** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8773/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4599/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7175/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8697/)

**Comment by prajakta-gokhale** [comment]
MacOS CI rebuild:

* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7184/)

**Pull request merged**